### PR TITLE
Port to FFmpeg 5.x

### LINKED ISSUE #188: Please add support for ffmpeg 5
Hi,
````
Linux (Debian unstable)
amd64
GCC 11.3.0
ffmpeg 5.0
````
Deprecated functions introduced in ffmpeg 4 have been removed and thus inputstream.ffmpegdirect
fail to build with ffmpeg 5.0
```
[1/3] Building CXX object CMakeFiles/inputstream.ffmpegdirect.dir/src/stream/FFmpegStream.cpp.o
FAILED: CMakeFiles/inputstream.ffmpegdirect.dir/src/stream/FFmpegStream.cpp.o 
/usr/lib/ccache/c++ -DADDON_GLOBAL_VERSION_FILESYSTEM_USED -DADDON_GLOBAL_VERSION_GENERAL_USED -DADDON_GLOBAL_VERSION_MAIN_USED -DADDON_GLOBAL_VERSION_NETWORK_USED -DADDON_GLOBAL_VERSION_TOOLS_USED -DADDON_INSTANCE_VERSION_INPUTSTREAM_USED -DBUILD_KODI_ADDON -DFFMPEGDIRECT_VERSION=19.0.3 -Dinputstream_ffmpegdirect_EXPORTS  -g -O2 -ffile-prefix-map=/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -O2 -g -DNDEBUG -fPIC   -DTARGET_POSIX -DTARGET_LINUX -D_GNU_SOURCE -DHAVE_LINUX_UDMABUF=1 -DHAVE_LINUX_DMA_HEAP=1 -DHAVE_LINUX_DMA_BUF=1 -DHAVE_MKOSTEMP=1 -DHAVE_LINUX_MEMFD=1 -DHAVE_STATX=1 -DHAVE_SSE=1 -DHAVE_SSE2=1 -DHAVE_SSE3=1 -DHAVE_SSSE3=1 -DHAVE_SSE4_1=1 -std=c++14 -MD -MT CMakeFiles/inputstream.ffmpegdirect.dir/src/stream/FFmpegStream.cpp.o -MF CMakeFiles/inputstream.ffmpegdirect.dir/src/stream/FFmpegStream.cpp.o.d -o CMakeFiles/inputstream.ffmpegdirect.dir/src/stream/FFmpegStream.cpp.o -c /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp
In file included from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.h:13,
                 from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:8:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/DemuxStream.h:36:7: warning: ‘ffmpegdirect::DemuxStream’ declared with greater visibility than the type of its field ‘ffmpegdirect::DemuxStream::cryptoSession’ [-Wattributes]
   36 | class DemuxStream
      |       ^~~~~~~~~~~
In file included from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.h:14,
                 from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:8:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/CurlInput.h:23:7: warning: ‘ffmpegdirect::CurlInput’ declared with greater visibility than the type of its field ‘ffmpegdirect::CurlInput::m_pFile’ [-Wattributes]
   23 | class CurlInput
      |       ^~~~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘bool ffmpegdirect::FFmpegStream::Open(bool)’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:684:37: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
  684 |       iformat = av_find_input_format("mpeg");
      |                 ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
      |                                     |
      |                                     const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:686:37: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
  686 |       iformat = av_find_input_format("mpeg");
      |                 ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
      |                                     |
      |                                     const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:688:37: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
  688 |       iformat = av_find_input_format("mpegts");
      |                 ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
      |                                     |
      |                                     const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:690:37: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
  690 |       iformat = av_find_input_format("mjpeg");
      |                 ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
      |                                     |
      |                                     const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:737:40: error: assignment of member ‘AVInputFormat::flags’ in read-only object
  737 |       m_pFormatContext->iformat->flags |= AVFMT_NOGENSEARCH;
      |                                        ^
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:789:30: error: ‘AVFMT_FLAG_KEEP_SIDE_DATA’ was not declared in this scope
  789 |   m_pFormatContext->flags |= AVFMT_FLAG_KEEP_SIDE_DATA;
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:831:72: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘codec_info_nb_frames’
  831 |             if ((st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO && st->codec_info_nb_frames > 0) ||
      |                                                                        ^~~~~~~~~~~~~~~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘bool ffmpegdirect::FFmpegStream::OpenWithCURL(AVInputFormat*)’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:994:35: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
  994 |     iformat = av_find_input_format("s16be");
      |               ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
      |                                   |
      |                                   const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1002:42: error: invalid conversion from ‘AVInputFormat**’ to ‘const AVInputFormat**’ [-fpermissive]
 1002 |       av_probe_input_buffer(m_ioContext, &iformat, strFile.c_str(), NULL, 0, 0);
      |                                          ^~~~~~~~
      |                                          |
      |                                          AVInputFormat**
In file included from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/DemuxStream.h:25,
                 from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.h:13,
                 from /home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:8:
/usr/include/x86_64-linux-gnu/libavformat/avformat.h:2070:66: note:   initializing argument 2 of ‘int av_probe_input_buffer(AVIOContext*, const AVInputFormat**, const char*, void*, unsigned int, unsigned int)’
 2070 | int av_probe_input_buffer(AVIOContext *pb, const AVInputFormat **fmt,
      |                                            ~~~~~~~~~~~~~~~~~~~~~~^~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1043:40: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
 1043 |         iformat2 = av_find_input_format("spdif");
      |                    ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
      |                                        |
      |                                        const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1052:42: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
 1052 |           iformat2 = av_find_input_format("dts");
      |                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~
      |                                          |
      |                                          const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1075:39: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
 1075 |         iformat = av_find_input_format("aac");
      |                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
      |                                       |
      |                                       const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1077:39: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
 1077 |         iformat = av_find_input_format("aac");
      |                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
      |                                       |
      |                                       const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1079:39: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
 1079 |         iformat = av_find_input_format("flv");
      |                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
      |                                       |
      |                                       const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1081:39: error: invalid conversion from ‘const AVInputFormat*’ to ‘AVInputFormat*’ [-fpermissive]
 1081 |         iformat = av_find_input_format("flv");
      |                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~
      |                                       |
      |                                       const AVInputFormat*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘virtual void ffmpegdirect::FFmpegStream::UpdateCurrentPTS()’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1150:27: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘cur_dts’
 1150 |     if (stream && stream->cur_dts != (int64_t)AV_NOPTS_VALUE)
      |                           ^~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1152:44: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘cur_dts’
 1152 |       double ts = ConvertTimestamp(stream->cur_dts, stream->time_base.den, stream->time_base.num);
      |                                            ^~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘void ffmpegdirect::FFmpegStream::ParsePacket(AVPacket*)’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1567:44: error: invalid conversion from ‘const AVCodec*’ to ‘AVCodec*’ [-fpermissive]
 1567 |       AVCodec* codec = avcodec_find_decoder(st->codecpar->codec_id);
      |                        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
      |                                            |
      |                                            const AVCodec*
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘ffmpegdirect::TRANSPORT_STREAM_STATE ffmpegdirect::FFmpegStream::TransportStreamAudioState()’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1644:42: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘cur_dts’
 1644 |             m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
      |                                          ^~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1664:42: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘cur_dts’
 1664 |             m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
      |                                          ^~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘ffmpegdirect::TRANSPORT_STREAM_STATE ffmpegdirect::FFmpegStream::TransportStreamVideoState()’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1697:42: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘cur_dts’
 1697 |             m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
      |                                          ^~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1717:42: error: ‘AVStream’ {aka ‘struct AVStream’} has no member named ‘cur_dts’
 1717 |             m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
      |                                          ^~~~~~~
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘ffmpegdirect::DemuxStream* ffmpegdirect::FFmpegStream::AddStream(int)’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:1834:9: error: ‘av_get_channel_layout_string’ was not declared in this scope; did you mean ‘av_get_sample_fmt_string’?
 1834 |         av_get_channel_layout_string(buf, 31, st->iChannels, st->iChannelLayout);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         av_get_sample_fmt_string
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp: In member function ‘virtual std::string ffmpegdirect::FFmpegStream::GetStreamCodecName(int)’:
/home/marillat/src/kodi-inputstream-ffmpegdirect-19.0.3/src/stream/FFmpegStream.cpp:2073:42: error: invalid conversion from ‘const AVCodec*’ to ‘AVCodec*’ [-fpermissive]
 2073 |     AVCodec* codec = avcodec_find_decoder(stream->codec);
      |                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
      |                                          |
      |                                          const AVCodec*
ninja: build stopped: subcommand failed.
```

### PR DESCRIPTION
**Pull request by basilgello**
Fixes #188

### CODE REVIEW

File: src/stream/FFmpegStream.cpp [hunk 1]
```diff
@@ -1572,6 +1573,146 @@ bool FFmpegStream::SeekTime(double time, bool backwards, double* startpts)
     return false;
 }
 
+int FFmpegStream::GetPacketExtradata(const AVPacket* pkt, const AVCodecParserContext* parserCtx, AVCodecContext* codecCtx, uint8_t **p_extradata)
```
Comment (phunkyfish) [inline src/stream/FFmpegStream.cpp:1546]: Nice work, and also a nice encapsulation of for the extra data.

Comment (basilgello) [reply src/stream/FFmpegStream.cpp:1546]: Thanks! :) I will push it once more following the const fixes in core PR to make builds happy with 4.4

File: src/stream/FFmpegStream.cpp [hunk 2]
```diff
@@ -729,14 +732,6 @@ bool FFmpegStream::Open(bool fileinfo)
...
   }
 
-  if (iformat && (strcmp(iformat->name, "mov,mp4,m4a,3gp,3g2,mj2") == 0))
-  {
-    CURL url(m_streamUrl);
-    //if (URIUtils::IsRemote(strFile))
-    if (!url.GetProtocol().empty() && !url.IsProtocol("file"))
-      m_pFormatContext->iformat->flags |= AVFMT_NOGENSEARCH;
-  }
-
```
Comment (phunkyfish) [inline src/stream/FFmpegStream.cpp:739]: Why remove this?

Comment (basilgello) [reply src/stream/FFmpegStream.cpp:739]: @AlwinEsch did this in core PR https://github.com/xbmc/xbmc/pull/21248/commits/aec31f55115df5d62905688d0cd2dbe3ccfab1ac#diff-bad369057589a59119615d8f380bff839b51c2100f2e8b5fd92b24fb61ea3fc6R546

Comment (phunkyfish) [reply src/stream/FFmpegStream.cpp:739]: @AlwinEsch do you recall the reason for this?

**Comment by herrnst** [comment]
Hi, not sure if its okay to already ask questions as even the FFMPEG 5 pull isn't merged into the Kodi repository yet, I'll try anyways:

I picked up current Kodi master (which will become Nexus), applied https://github.com/xbmc/xbmc/pull/21248 and built/installed everything (worked without issues so far). This also installed the ffmpeg-related .a archives in /usr/lib/x864_64-linux-gnu/kodi/ffmpeg/lib/lib*.a and the headers below /usr/include/kodi/lib* (ie. libavcodec et al).

Then I picked up this PR branch (inputstream.ffmpegdirect+ffmpeg5) and invoked CMake with
`-DCMAKE_PREFIX_PATH=/usr/lib/x864_64-linux-gnu/kodi/ffmpeg` and CMake picks up FFmpeg from Kodi.

Compilation works, but at the end linking fails with
```
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/kodi/ffmpeg/lib/libavcodec.a(h264_intrapred_10bit.o): warning: relocation against `ff_pw_512' in read-only section `.text'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/kodi/ffmpeg/lib/libavcodec.a(vc1dsp_mmx.o): relocation R_X86_64_PC32 against symbol `ff_pw_9' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```

This is on a minimal Ubuntu 22.04 install (used solely for packaging purposes) with all current updates installed. What am I doing wrong?

**Comment by basilgello** [comment]
@herrnst How do you build? You build dsc from ppa?

**Comment by herrnst** [comment]
> @herrnst How do you build? You build dsc from ppa?

Not sure if the answer is "yes", basically I pick up packaging ("debian/*") from xbmc-nightly at launchpad.net/~team-xbmc (https://launchpad.net/~team-xbmc/+archive/ubuntu/xbmc-nightly/+sourcepub/14089652/+listing-archive-extra) and use that to build from the GIT checkout (here with FFMPEG 5 patches applied) using dpkg-buildpackage with all deps installed. Same for inputstream.ffmpegdirect, so thats where kodi-ffmpeg-dev (ie. /usr/lib/*/libavcodec.a et al) come from.

So, short answer: debian-dir from PPA with GIT tree and debhelper/dh-make and friends, both for Kodi and inputstream.ffmpegdirect.

**Comment by basilgello** [comment]
@phunkyfish The ffmpeg5 PR is merged into master. Can you please create Omega branch so I can rebase the PR into it? As I told previously, we need to use the same ffmpeg version across kodi and addons.

**Comment by phunkyfish** [comment]
Creating a new branch won’t work until all binary add-ons have branches switched from Nexus to Omega (but if you ask me master would be better).

I’m not sure we can branch for a single add-on and leave the others as is.

**Comment by basilgello** [comment]
I am afraid that the ffmpeg5 upgrade on kodi core will break ffmpegdirect earlier built with bundled ffmpeg4 for most users. Windows users might not notice that as well as Debian's, but other Linux / OSX / Android / FreeBSD certainly will :(

As a workaround until the Omega repo goes prime time users can install the artifacts from this PR. In this case this PR should be held on until we have Omega branch.

**Comment by phunkyfish** [comment]
When ffmpeg 4 was released it was out for many months on ffmpegdirect before being in kodi. If it worked then, it could also work now with ffmpeg 5. But the only way to see if it does is if we try it.

**Comment by lrusak** [comment]
Omega is currently pre-alpha so I wouldn't worry about breakage too much.

**Comment by phunkyfish** [comment]
The issue is we could break Nexus as the same version of the add-on would be used for both.

But let’s test both first and see what happens.

**Comment by basilgello** [comment]
OK, did some tests with team's PPA for jammy:

1. kodi 6:20.0+git20230105.0300-10f29c9fd4-0~jammy (ffmpeg4) + ffmpegdirect 20.5.0 with this PR (ffmpeg5) = works!
2. kodi 6:20.0+git20230107.0300-8fdbdb1890-0~jammy (ffmpeg5) + ffmpegdirect 20.5.0 from PPA (ffmpeg4) = no picture and symptoms are just like the extradata

So we can merge this PR - and the faster we merge it, the less chance users will get combo number 2 :)

**Comment by basilgello** [comment]
And if we still support Matrix branch, the use-after-free commit likely applies to matrix as well…

**Comment by phunkyfish** [comment]
> And if we still support Matrix branch, the use-after-free commit likely applies to matrix as well…

Please PR that change also and I’ll merge it.

**Comment by lrusak** [comment]
> The issue is we could break Nexus as the same version of the add-on would be used for both.
> 
> 
> 
> But let’s test both first and see what happens.

Ah, why bump to 5.0 for the Nexus version when Kodi Nexus is still 4.x? Is it not valid to keep them in sync?

**Comment by basilgello** [comment]
That was my concern, too :) But things seem to work if core ffmpeg is older than addon's - if the linkage of the addon is static and there are no other binary addons trying to use dynamic ffmpeg library!

**Comment by lrusak** [comment]
> That was my concern, too :) But things seem to work if core ffmpeg is older than addon's - if the linkage of the addon is static and there are no other binary addons trying to use dynamic ffmpeg library!

Yea that's annoying. Now LibreELEC (and other distros) need to manage two ffmpeg versions and build them statically.

**Comment by basilgello** [comment]
The alternative is to do quick omega branching and revert this commit :) I can help you with it tonight. And please upload,tzdata to mirrors, too!

**Comment by phunkyfish** [comment]
> > That was my concern, too :) But things seem to work if core ffmpeg is older than addon's - if the linkage of the addon is static and there are no other binary addons trying to use dynamic ffmpeg library!
> 
> Yea that's annoying. Now LibreELEC (and other distros) need to manage two ffmpeg versions and build them statically.

Why? Ffmpegdirect should be able to change the underlying ffmpeg version without impacting kodi.

**Comment by lrusak** [comment]
> > > That was my concern, too :) But things seem to work if core ffmpeg is older than addon's - if the linkage of the addon is static and there are no other binary addons trying to use dynamic ffmpeg library!
> 
> > 
> 
> > Yea that's annoying. Now LibreELEC (and other distros) need to manage two ffmpeg versions and build them statically.
> 
> 
> 
> Why? Ffmpegdirect should be able to change the underlying ffmpeg version without impacting kodi. 

Until we allow using whatever ffmpeg version available in Kodi core distros will have to manage two ffmpeg versions.

I agree that the add-on should not need the exact version of ffmpeg that core uses but we aren't quite there yet.

**Comment by HiassofT** [comment]
Just a quick heads up: if you start to require a different ffmpeg version than the fork kodi uses (eg 4.x for Nexus) then we (LE) might just choose to not build the ffmpegdirect addon any longer as  it'll become too much hassle.

Ideally both kodi and ffmpegdirect should work with non-forked upstreeam ffmpeg versions (ideally both 4.x and 5.x - your choice) as provided by distros - that should be the mid-term goal both for kodi omega and LE12

**Comment by basilgello** [comment]
@HiassofT I did everything to reduce hassle here :)

First of all, LE can build Nexus with ffmpeg5 (using the [commit from Debian](https://salsa.debian.org/multimedia-team/kodi-media-center/kodi/-/blob/master/debian/patches/workarounds/0002-ffmpeg5.patch))

Another option is to add the patch (reverting 60ffdc3df358d09540961d54a2b7f9e45670002a) [here](https://github.com/LibreELEC/LibreELEC.tv/tree/master/packages/mediacenter/kodi-binary-addons/inputstream.ffmpegdirect) to build ffmpegdirect with ffmpeg4 (whipe preserving the ability to build with ffmpeg5 at the same time!).

Finally, when Omega gets branched in addons, we can revert this PR as a whole for Nexus. That sounds a proper long-term solution to me.

**Comment by lrusak** [comment]
> Finally, when Omega gets branched in addons, we can revert this PR as a whole for Nexus. That sounds a proper long-term solution to me.

I think this is the best option and should happen sooner than later.

Omega can be branched at any time. Nothing is holding that back really.

**Comment by razzeee** [comment]
Why are we merging this to a nexus branch at all then? This is causing failing builds for no reason?

**Comment by phunkyfish** [comment]
Will revert later today.

**Comment by HiassofT** [comment]
@basilgello We are not going to patch kodi Nexus to work with ffmpeg5, this will create a huge maintenance burden.

I, too, find it unfortunate that team kodi decided to ship Nexus with ffmpeg 4 but it is what it is and we have to live with that.

Kodi doesn't sandbox binary addons (eg into spearate processes) yet so you can't easily mix libraries - kodi's address space will already have ffmpeg 4 symbols present and trying to mix that with ffmpeg 5 is crying for problems (and likely very sublte issues that are going to be hard to track down).

I fully understand that you'll want to ship kodi Nexus with ffmpeg 5 on Debian but this means you're going to have to locally patch kodi and addons to do that.

**Comment by phunkyfish** [comment]
Reverted. Once binary add-ons all get branched for Omega we can apply this PR again.

**Comment by basilgello** [comment]
I have no hard feelings about this but I recall there was a script posted by @fuzzard in slack to automate branching. What was the issue there?

**Comment by phunkyfish** [comment]
I did not see the script. But it will need someone with with the correct permissions to run it for all binary addons.

**Pull request merged**